File: lab1/src/T3.java
import java.util.Arrays;

public class T3 extends Thread {
    private int x3 = 0;

    public void run() {
        try {
            System.out.println("T3 is started");

            //Input B
            Arrays.fill(Data.B, 1);

            //Input p
            Data.p = 1;

            //Signal T1, T2, T4 about input B, p
            Data.Sem3.release(3);

            //Wait for input MC, E in T1
            Data.Sem1.acquire(1);

            //Wait for input MD, d in T2
            Data.Sem2.acquire(1);

            //Wait for input R, Z in T4
            Data.Sem4.acquire(1);

            //Calculation1 Sh = R * MCh
            Data.calculation1(Data.H * 2, Data.H * 3);

            //Calculation2 x3 = BH * ZH
            for (int i = 0; i < Data.H; i++) {
                x3 += Data.B[i] * Data.Z[i];
            }

            //Calculation3 x = x + x3
            Data.x.addAndGet(x3);

            //Signal T1, T2, T4 about end of calculation x
            Data.Sem7.release(3);

            //Wait for end of calculation x in T1
            Data.Sem5.acquire(1);

            //Wait for end of calculation x in T2
            Data.Sem6.acquire(1);

            //Wait for end of calculation x in T4
            Data.Sem8.acquire(1);

            //Copy p3 = p
            int p3;
            synchronized (Data.Y) {
                p3 = Data.p;
            }

            //Copy x3 = x
            Data.Sem0.acquire();
            x3 = Data.x.get();
            Data.Sem0.release();

            //Copy d3 = d
            int d3;
            synchronized (Data.W) {
                d3 = Data.d;
            }

            //Calculation4 Ah = S * MDh * p3 + x3 * Eh * d3
            Data.calculation4(p3, x3, d3, 0, Data.H);

            //Wait for end of calculation Ah in T1
            Data.Sem9.acquire(1);

            //Wait for end of calculation Ah in T2
            Data.Sem10.acquire(1);

            //Wait for end of calculation Ah in T4
            Data.Sem11.acquire(1);

            //Output Ah
            System.out.println(Arrays.toString(Data.A));

            System.out.println("T3 is finished");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: lab2/ConsoleApp1/T3.cs
namespace ConsoleApp1;

public class T3
{
    public static void Run(Data data)
    {
        Console.WriteLine("T3 is started");

        // Wait for input B, MX in T2
        data.Sem1.WaitOne();

        // Wait for input d, Z, MM in T4
        data.Sem2.WaitOne();

        data.Mutex1.WaitOne();
        // Copy d3 = d
        var d3 = data.d;
        data.Mutex1.ReleaseMutex();

        // Calculation1 Ah = sort(d * Bh + Z * (MM * MXh))
        data.calculation1(d3, data.H * 2, data.H * 3);

        // Wait for calculation Ah in T4
        data.Sem5.WaitOne();

        // Calculation2 A2h = sort(Ah, Ah)
        data.calculation2(data.H * 2, data.H * 4);

        // Signal T1 about end of calculation A2h
        data.Sem4.Release();

        // Calculation4 ai = min(Bh)
        var a3 = data.calculation4(data.H * 2, data.H * 3);

        // Calculation5 a = min(a, ai)
        lock(data.aLock)
        {
            data.a = Math.Min(data.a, a3);
        }

        // Signal T1, T2, T4 about end of calculation a
        data.Event3.Set();

        // Wait for calculation a, A in T1
        data.Event1.WaitOne();

        // Wait for calculation a in T2
        data.Event2.WaitOne();

        // Wait for calculation a in T4
        data.Event4.WaitOne();

        data.Mutex2.WaitOne();
        // Copy a3 = a
        a3 = data.a;
        data.Mutex2.ReleaseMutex();

        // Calculation6 Xh = Ah * a
        data.calculation6(a3, data.H * 2, data.H * 3);

        // Signal T4 about end of calculation Xh
        data.Barrier.SignalAndWait();

        Console.WriteLine("T3 is finished");
    }
}
